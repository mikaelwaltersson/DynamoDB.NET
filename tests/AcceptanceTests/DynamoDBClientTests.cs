using Amazon.DynamoDBv2;
using DynamoDB.Net;
using DynamoDB.Net.Model;
using DynamoDB.Net.Serialization;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Snapshooter.Xunit;
using Xunit;

namespace DynamoDB.NET.Tests.AcceptanceTests
{
    public partial class DynamoDBClientTests
    {
        [Fact]
        public async Task CanCreateTables()
        {
 
            var services = new ServiceCollection();

            services.AddOptions();
            services.AddLogging();
            
            services.AddSingleton(
                typeof(IAmazonDynamoDB),
                serviceProvider =>
                    new AmazonDynamoDBClient(
                        new AmazonDynamoDBConfig
                        {
                            ServiceURL = "http://localhost:8000"
                        }));

            services.AddSingleton<IDynamoDBClient, DynamoDBClient>();
            services.Configure<DynamoDBClientOptions>(
                options => 
                {
                    options.TableNamePrefix = "acceptance-tests-";
                }
            );

            var serviceProvider = services.BuildServiceProvider();

            var dynamoDb = serviceProvider.GetRequiredService<IAmazonDynamoDB>();
            var dynamoDbClientOptions = serviceProvider.GetRequiredService<IOptions<DynamoDBClientOptions>>();
            var client = serviceProvider.GetRequiredService<IDynamoDBClient>();

            var table = TableDescription.Get(typeof(TestModels.UserPost), JsonContractResolver.DefaultDynamoDB);
            var createTableRequest = table.GetCreateTableRequest(dynamoDbClientOptions.Value);
            
            var result = await dynamoDb.CreateTableAsync(createTableRequest);
            try 
            {
                Assert.Equal("acceptance-tests-user-posts", result.TableDescription.TableName);

                var item1 = await client.PutAsync(
                    new TestModels.UserPost 
                    { 
                        UserId = new Guid("00000000-0000-0000-0000-000000000001"), 
                        Timestamp = new DateTime(2022, 10, 18, 16, 42, 0),
                        RoleIds =
                        {
                            new Guid("00000000-0000-0000-0000-000000000002"), 
                            new Guid("00000000-0000-0000-0000-000000000003")
                        }
                    });
                var item2 = await client.TryGetAsync(new PrimaryKey<TestModels.UserPost>(new Guid("00000000-0000-0000-0000-000000000001"), new DateTime(2022, 10, 18, 16, 42, 0)));

                Assert.NotNull(item2);
                Snapshot.Match(item1);
            }
            finally 
            {
                await dynamoDb.DeleteTableAsync(result.TableDescription.TableName);
            }
        }
    }
}