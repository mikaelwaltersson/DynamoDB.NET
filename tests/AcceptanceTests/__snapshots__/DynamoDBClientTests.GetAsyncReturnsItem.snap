{
  "UserId": "f4ebd04f-0bd7-43b9-90a8-d6de295927f3",
  "Timestamp": "2022-10-18T16:42:00Z",
  "RoleIds": [
    "aa5988f8-4a3f-40ae-987b-22bac870b170",
    "14598640-6c86-4206-b1a6-753e640b008a"
  ]
}
